vaultclient/vaultclient.go

```diff
@@ -1,63 +1,177 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign request: %w", err)
 	}
 
-	// Handlers
-	svc.Handlers.Sign.PushBackNamed(v4.SignRequestHandler)
-	svc.Handlers.Build.PushBackNamed(query.BuildHandler)
-	svc.Handlers.Unmarshal.PushBackNamed(restjson.UnmarshalHandler)
-	svc.Handlers.UnmarshalError.PushBackNamed(query.UnmarshalErrorHandler)
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to make request: %w", err)
+	}
+
+	return resp, nil
+}
 
-	return svc
+type APIError struct {
+	StatusCode int
+	Err        error
 }
 
-// newRequest creates a new request for a Vault operation and runs any
-// custom request initialization.
-func (c *Vault) newRequest(op *request.Operation, params, data interface{}) *request.Request {
-	req := c.NewRequest(op, params, data)
+func (h *APIError) Error() string {
+	return h.Err.Error()
+}
+
+func (h *APIError) Unwrap() error {
+	return h.Err
+}
+
+func (h *APIError) HTTPStatusCode() int {
+	return h.StatusCode
+}
+
+func (vc *Vault) handleAWSResponse(resp *http.Response) (respBody []byte, err error) {
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil && err == nil {
+			err = fmt.Errorf("failed to close response body: %w", closeErr)
+		}
+	}()
+
+	respBody, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response: %w", err)
+	}
+
+	if resp.StatusCode >= 400 {
+		var xmlError struct {
+			Error struct {
+				Code    string `xml:"Code"`
+				Message string `xml:"Message"`
+			} `xml:"Error"`
+			RequestID string `xml:"RequestId"`
+		}
+
+		if err := xml.Unmarshal(respBody, &xmlError); err != nil {
+			message := string(respBody)
+			return nil, &APIError{
+				StatusCode: resp.StatusCode,
+				Err: &types.ServiceFailureException{
+					Message: &message,
+				},
+			}
+		}
+
+		switch xmlError.Error.Code {
+		case (&types.EntityAlreadyExistsException{}).ErrorCode():
+			return nil, &APIError{
+				StatusCode: resp.StatusCode,
+				Err: &types.EntityAlreadyExistsException{
+					Message: &xmlError.Error.Message,
+				},
+			}
+		case (&types.InvalidInputException{}).ErrorCode():
+			return nil, &APIError{
+				StatusCode: resp.StatusCode,
+				Err: &types.InvalidInputException{
+					Message: &xmlError.Error.Message,
+				},
+			}
+		case (&types.NoSuchEntityException{}).ErrorCode():
+			return nil, &APIError{
+				StatusCode: resp.StatusCode,
+				Err: &types.NoSuchEntityException{
+					Message: &xmlError.Error.Message,
+				},
+			}
+		case (&types.DeleteConflictException{}).ErrorCode():
+			return nil, &APIError{
+				StatusCode: resp.StatusCode,
+				Err: &types.DeleteConflictException{
+					Message: &xmlError.Error.Message,
+				},
+			}
+		default:
+			return nil, &APIError{
+				StatusCode: resp.StatusCode,
+				Err: &types.ServiceFailureException{
+					Message: &xmlError.Error.Message,
+				},
+			}
+		}
+
+	}
 
-	return req
+	return respBody, nil
```

Comment:
nit: could do the unmarshalling here, to simplify code in every use

```go
func (vc *Vault) handleAWSResponse(resp *http.Response, result interface{}) err error 
{
    [...]

    return json.Unmarshall(respBody, result)
}
```

(i wish we could result both `result` and `error`, but this would require generics...which are not supported on methods)